Mathlib/RingTheory/GradedAlgebra/Homogeneous/Submodule.lean

```diff
@@ -0,0 +1,108 @@
+/-
+Copyright (c) 2021 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Eric Wieser
+-/
+import Mathlib.RingTheory.GradedAlgebra.Basic
+import Mathlib.Algebra.GradedMulAction
+
+/-!
+# Homogeneous submodules of a graded module
+
+This file defines homogeneous submodule of a graded module `⨁ᵢ ℳᵢ` over graded ring `⨁ᵢ 𝒜ᵢ` and
+operations on them.
+
+## Main definitions
+
+For any `p : Submodule A M`:
+* `Submodule.IsHomogeneous ℳ p`: The property that a submodule is closed under `GradedModule.proj`.
+* `HomogeneousSubmodule 𝒜 ℳ`: The structure extending submodules which satisfy
+  `Submodule.IsHomogeneous`.
+
+## Implementation notes
+
+The **notion** of homogeneous submodule does not rely on a graded ring, only a decomposition of the
+the module. However, most interesting properties of homogeneous submodules do rely on the base ring
+being a graded ring. For technical reasons, we make `HomogeneousSubmodule` depend on a graded ring.
+For example, if the definition of a homogeneous submodule does not depend on a graded ring, the
+instance that `HomogeneousSubmodule` is a complete lattice cannot be synthesized due to
+synthesation order.
+
+## Tags
+
+graded algebra, homogeneous
+-/
+
+open SetLike DirectSum Pointwise Set
+
+variable {ιA ιM σA σM A M : Type*}
+
+variable [Semiring A] [AddCommMonoid M] [Module A M]
+
+section HomogeneousDef
+
+/-- An `p : Submodule A M` is homogeneous if for every `m ∈ p`, all homogeneous components
+  of `m` are in `p`. -/
+def Submodule.IsHomogeneous (p : Submodule A M) (ℳ : ιM → σM)
+    [DecidableEq ιM] [SetLike σM M] [AddSubmonoidClass σM M] [Decomposition ℳ] : Prop :=
+  ∀ (i : ιM) ⦃m : M⦄, m ∈ p → (DirectSum.decompose ℳ m i : M) ∈ p
```

Comment:
I wonder if this makes sense as
```suggestion
def AddSubmonoidClass.IsHomogeneous (p : σM) (ℳ : ιM → σM)
    [DecidableEq ιM] [SetLike σM M] [AddSubmonoidClass σM M] [Decomposition ℳ] : Prop :=
  ∀ (i : ιM) ⦃m : M⦄, m ∈ p → (DirectSum.decompose ℳ m i : M) ∈ p
```

Reply:
Let me try

Reply:
Yes, it's fine to specialize `HomogeneousSubmodule`

Reply:
But then the downstream `Ideal.IsHomogeneous` need to be changed as well.
```lean

variable [Semiring A]
variable [SetLike σ A] [AddSubmonoidClass σ A] (𝒜 : ι → σ)
variable [DecidableEq ι] [AddMonoid ι] [GradedRing 𝒜]
variable (I : Ideal A)

/-- An `I : Ideal A` is homogeneous if for every `r ∈ I`, all homogeneous components
  of `r` are in `I`. -/
def Ideal.IsHomogeneous : Prop := AddSubmonoidClass.IsHomogeneous 𝒜 I
```
for now `Ideal` is not `σ`. Not sure if this is direction we want

Reply:
```lean
/-- An `p : Submodule A M` is homogeneous if for every `m ∈ p`, all homogeneous components
  of `m` are in `p`. -/
def AddSubmonoidClass.IsHomogeneous {P : Type*} [SetLike P M] [AddSubmonoidClass P M]
    (ℳ : ιM → σM) (p : P)
    [DecidableEq ιM] [SetLike σM M] [AddSubmonoidClass σM M] [Decomposition ℳ] : Prop :=
  ∀ (i : ιM) ⦃m : M⦄, m ∈ p → (DirectSum.decompose ℳ m i : M) ∈ p
```
Maybe this is better?

Reply:
I will open a separate PR for `SetLike.IsHomogeneous` and `SetLike.isHomogeneousElement`